core/src/services/supabase/backend.rs

```diff
@@ -0,0 +1,282 @@
+            Some(endpoint.trim_end_matches('/').to_string())
+        };
+
+        self
+    }
+
+    /// Set the authorization key for this backend
+    /// Do not set this key if you want to read public bucket
+    pub fn key(&mut self, key: &str) -> &mut Self {
+        self.key = Some(key.to_string());
+        self
+    }
+
+    /// Specify the http client that used by this service.
+    ///
+    /// # Notes
+    ///
+    /// This API is part of OpenDAL's Raw API. `HttpClient` could be changed
+    /// during minor updates.
+    pub fn http_client(&mut self, client: HttpClient) -> &mut Self {
+        self.http_client = Some(client);
+        self
+    }
+}
+
+impl Builder for SupabaseBuilder {
+    const SCHEME: Scheme = Scheme::Supabase;
+    type Accessor = SupabaseBackend;
+
+    fn from_map(map: std::collections::HashMap<String, String>) -> Self {
+        let mut builder = SupabaseBuilder::default();
+
+        map.get("root").map(|v| builder.root(v));
+        map.get("bucket").map(|v| builder.bucket(v));
+        map.get("endpoint").map(|v| builder.endpoint(v));
+        map.get("key").map(|v| builder.key(v));
+
+        builder
+    }
+
+    fn build(&mut self) -> Result<Self::Accessor> {
+        let root = normalize_root(&self.root.take().unwrap_or_default());
+        debug!("backend use root {}", &root);
+
+        let bucket = &self.bucket;
+
+        let endpoint = self.endpoint.take().unwrap_or_default();
+
+        let http_client = if let Some(client) = self.http_client.take() {
+            client
+        } else {
+            HttpClient::new().map_err(|err| {
+                err.with_operation("Builder::build")
+                    .with_context("service", Scheme::Supabase)
+            })?
+        };
+
+        let key = self.key.as_ref().map(|k| k.to_owned());
+
+        let core = SupabaseCore::new(&root, bucket, &endpoint, key, http_client);
+
+        let core = Arc::new(core);
+
+        Ok(SupabaseBackend { core })
+    }
+}
+
+#[derive(Debug)]
+pub struct SupabaseBackend {
+    core: Arc<SupabaseCore>,
+}
+
+#[async_trait]
+impl Accessor for SupabaseBackend {
+    type Reader = IncomingAsyncBody;
+    type BlockingReader = ();
+    type Writer = SupabaseWriter;
+    type BlockingWriter = ();
+    // todo: implement Pager to support list and scan
+    type Pager = ();
+    type BlockingPager = ();
+
+    fn info(&self) -> AccessorInfo {
+        let mut am = AccessorInfo::default();
+        am.set_scheme(Scheme::Supabase)
+            .set_root(&self.core.root)
+            .set_name(&self.core.bucket)
+            .set_capability(Capability {
+                stat: true,
+                read: true,
+                write: true,
+
+                ..Default::default()
+            });
+
+        am
+    }
+
+    async fn read(&self, path: &str, _args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let resp = if self.core.key.is_some() {
+            self.core.supabase_get_object_auth(path).await?
```

Comment:
We only need to have `supabase_get_object`

Reply:
Got it